### DISCUSSION TITLE
NodeJS / Express / TypeScript Deployment with dist directory

### DISCUSSION THREAD

**Discussion by gitananun** [vercel/community#6389] [Help]
### Summary

Hi

I'm trying to deploy NodeJS / Express API application, but having issues with running the project.
Deployment is fine, but the URL opens the dist/index.js file instead of executing it.

Here is my `vercel.json` file:
```
{
  "version": 2,
  "buildCommand": "npm run build",
  "outputDirectory": "dist",
  "routes": [
    {
      "src": "/(.*)",
      "dest": "dist/index.js"
    }
  ]
}
```

And this is part of my `package.json` file:
```
...
"scripts": {
    "start": "nodemon",
    "clean": "rm -rf build",
    "build": "echo '✅ Building project...' && npm run clean && tsc && tsc-alias",
    "ts.check": "tsc --project tsconfig.json",
     ...
```

I've checked the dist folder, and everything is correct there. Output is good.

I've done the deployment without any issues when I wasn't using the typescript paths. After that, I have to run tsc-alias command, so I've changed the approach in vercel.json.

So, before I was just using vercel.json:
```
{
  "version": 2,
  "builds": [
    {
      "src": "src/index.ts",
      "use": "@vercel/node"
    }
  ],
  "routes": [
    {
      "src": "/(.*)",
      "dest": "src/index.ts"
    }
  ]
}
```
So, if there is a way to fix @ paths of typescript with this approach, I don't need to move to the dist approach.

### Steps to Reproduce

Use package.json build command, and vercel.json file configs.
Vercel is showing the sourcecode of index.js instead of executing.

**Comment by amyegan** [comment]
Hey @gitananun. We recommend that you follow this guide to [adapt your Express.js app](https://vercel.com/guides/using-express-with-vercel#adapting-to-serverless-environments) to work in a serverless environment: 

[vercel.com/guides/using-express-with-vercel](https://vercel.com/guides/using-express-with-vercel)

**Reply by gitananun** [reply]
Thanks @amyegan!

Following the guide, all is working fine, besides the paths for TS.
It continues to throw:
```
Cannot find module '@middlewares/index' Require stack: ...
```

**Comment by fahimuddin-brillmark** [comment]
@gitananun Any update on this? I am just stuck on deploying express app with dist build. If you found any solution, please post here.

**Reply by gitananun** [reply]
@fahimuddin-brillmark nothing worked on that side. Just continued working with relative paths.

**Comment by robinoch** [comment]
@gitananun @fahimuddin-brillmark 
@amyegan: thanks for unlocking! ☺️ 

I was able to use TypeScript paths on Vercel. The following solution uses `resolve-tspaths`.
The relevant files look like this (all located in project root):

package.json
```
...
  "scripts": {
    ...
    "build:vercel": "tsc -p tsconfig.json && resolve-tspaths -p tsconfig.json -s ./src -o ./src --ext ts",
    ...
  },
...
```

tsconfig.json
```
{
  ...
  "compilerOptions": {
    ...
    "outDir": "./dist",
    "baseUrl": ".",
    "paths": {
      "@/*": ["./src/*"]
    }
    ...
  }
}
```

vercel.json
```
{
  "outputDirectory": "dist"
  "buildCommand": "npm run build:vercel",
  ...
}
```

### Important factors
- leave the build directory as it is (e.g. "./dist")
- set output of `resolve-tspaths` to `./src`
- configure `resolve-tspaths` to replace in *.ts files (not *.js)

### Further information
- It also worked with `tscpaths` as it by default outputs to *.ts and *.js files:
  `"build:vercel": "tsc -p tsconfig.json && tscpaths -p tsconfig.json -s ./src -o ./src --ext ts"`
- As by the available options of `tsc-alias`, it should also be possible with something like this:
  `"build:vercel": "tsc -p tsconfig.json && tsc-alias -p tsconfig.json --outDir ./src --fileExtensions ts"` <-- **not tested!**
- As this changes are vercel-specific, I added a separate build script (build:vercel) and kept the default version for builds anywhere else. You can configure this in the `vercel.json` as shown above. This way it can not be forgotten to be configured during vercel project setup.
- This solution was figured out and tested with a Fastify project 
  --> I'd appreciate any feedback on whether this is a generally working solution, thx!
- @amyegan I believe that any more in-depth documentation from Vercel on the underlying mechanics of serverless builds would be highly appreciated by the community
- never give up using TS paths! 😎 In total I spent a considerable amount of time for not letting them go - setup is king!

**Reply by amyegan** [reply]
Thanks for sharing the solution that worked for you!
